### ISSUE TITLE
Accept compareTo function in stream.substreamAt

### ISSUE THREAD

**Issue by nicolas-zozol** [d-plaindoux/parsec#45]
Here is the current implementation of subStream:

```
// Stream 'a => [Comparable 'a] -> number -> boolean
    subStreamAt(s, index){
        for (var i = 0; i < s.length; i++) {
            var value = this.get(i + index);
            if (!value.isSuccess() || value.success() !== s[i]) { // <=== compareTo
                return false;
            }
        }

        return true;
    }
```

Suppose we want to create `P.stringIgnoreCase("john doe")`. We could send a compareTo function to `subStream(s, index, [compareTo])`.

```
if (!value.isSuccess() || !compareTo(value.success(), s[i]) ) {
                return false;
            }

```

Is it a good idea ?

**Comment by d-plaindoux** [comment]
Yes and no ... I think the subStream material should be replaced now:

```
// int -> Parser (List 'a') a'
function subStream(length) {
    return any().occurrence(length);
}

// int -> Parser string char
function subString(length) {
    return subStream(length).map((s) => s.join(""));
}

// string -> Parser string char
function string(s) {
    return doTry(subString(s.lenght).filter((r) => r === s));
}
```

Then you can filter, compare a string with a given size.

**Comment by nicolas-zozol** [comment]
And now I understand more clearly what subStream does :)

**Comment by d-plaindoux** [comment]
With this approach the computation is efficient because the iteration is managed by only one function - occurence -  reducing the code complexity.

**Comment by nicolas-zozol** [comment]
Closed by Parsec simplify #47
